Analysis Archive Restore Image throws incorrect HTTP Response code

<!-- Thanks for trying Anchore Engine and filing an issue! To help ensure that we can quickly and efficiently process issues and get fixes please answer a few questions-->

**Is this a request for help?**: No

---

**Is this a BUG REPORT or a FEATURE REQUEST?** (choose one): Bug Report

<!--
If this is a BUG REPORT, please:
  - Fill in as much of the template below as possible so we can reproduce and understand the issue.

If this is a FEATURE REQUEST, please:
  - Describe the feature *in detail* as much as possible including possible cli and api behavior and how to test the feature.

In both cases, we'll probably ask follow-up questions to ensure it's well understood. If we cannot reproduce the issue or
questions go unanswered for too long, we may close the issue. If we're wrong, PLEASE feel free to re-open it and explain why.
-->

<!-- To get version info:
anchore-cli --version
docker exec <engine_container> pip show anchore-engine
-->
**Version of Anchore Engine and Anchore CLI if applicable**: 
```
samdacanay@SDACANAY ~ % a --version
anchore-cli, version 0.7.2
```


**What happened**: Tried to restore an image from the archive, but the image hadn't yet been deleted from the engine context. This was due to the fact that the image was archived via the CLI as well (which doesn't automatically delete the image from the engine context.




**What did you expect to happen**:  Originally expected that the restore would work (not realizing the image had to be manually deleted outside of the archive flow). Now, just expect a different HTTP response code. Instead of 500, should probably be a 409

<!--
Log output can be from the container stdout or /var/log/anchore/anchore-*.
The /var/log/anchore/anchore-* logs are service specific and can help identify which components is having the issue, so please check
there as well as the stdout from the container.
-->
**Anchore CLI terminal dump**:
```
samdacanay@SDACANAY ~ % a analysis-archive images add sha256:08ca4f2c45316cb0b79faf8f07625046dbbc16f6f140b0c59da03f1323ada8a4
Image Digest                                                                   Archive Status        Details
sha256:08ca4f2c45316cb0b79faf8f07625046dbbc16f6f140b0c59da03f1323ada8a4        archived              Completed successfully
samdacanay@SDACANAY ~ % a analysis-archive images restore sha256:08ca4f2c45316cb0b79faf8f07625046dbbc16f6f140b0c59da03f1323ada8a4
Error: failed post url=http://localdev-anchore-engine-catalog:8082/v1/images
HTTP Code: 500
Detail: {u'error_codes': []}
samdacanay@SDACANAY ~ % a --debug analysis-archive images restore sha256:08ca4f2c45316cb0b79faf8f07625046dbbc16f6f140b0c59da03f1323ada8a4
DEBUG:anchorecli.clients.apiexternal:As Account = None
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): localhost:8228
DEBUG:urllib3.connectionpool:http://localhost:8228 "GET /v1 HTTP/1.1" 200 5
DEBUG:anchorecli.clients.apiexternal:As Account = None
DEBUG:anchorecli.clients.apiexternal:POST url=http://localhost:8228/v1/images
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): localhost:8228
DEBUG:urllib3.connectionpool:http://localhost:8228 "POST /v1/images HTTP/1.1" 500 149
DEBUG:anchorecli.cli.utils:fetched httpcode from response: 500
Error: failed post url=http://localdev-anchore-engine-catalog:8082/v1/images
HTTP Code: 500
Detail: {u'error_codes': []}
```

**Any relevant log output from /var/log/anchore**:
```
[service:enterprise-apiext] 2020-08-27 21:21:45+0000 [_GenericHTTPChannelProtocol,25485,127.0.0.1] [PoolThread-twisted.internet.reactor-0] [anchore_engine.clients.services.internal/dispatch()] [ERROR] Failed client call to service catalog for url: http://localdev-anchore-engine-catalog:8082/v1/images. Response: {'httpcode': 500, 'anchore_error_raw': 'b\'"Conflict: Image already exists in system. No restore possible"\\n\'', 'anchore_error_json': 'Conflict: Image already exists in system. No restore possible'}
```


**What docker images are you using**: Enterprise, 2.4.0-rc1


**How to reproduce the issue**: Archive an image, try to restore it before deleting it from engine


**Anything else we need to know**: N/A

--- a/anchore_engine/common/helpers.py
+++ b/anchore_engine/common/helpers.py
@@ -29,19 +29,24 @@ def make_response_error(errmsg, in_httpcode=None, details=None):
 
     if isinstance(errmsg, Exception):
         if 'anchore_error_json' in errmsg.__dict__:
-            if {'message', 'httpcode', 'detail'}.issubset(set(errmsg.__dict__['anchore_error_json'])):
-                ret.update(errmsg.__dict__['anchore_error_json'])
+            # Try to load it as json
+            try:
+                err_json = json.loads(errmsg.__dict__.get('anchore_error_json', None))
+            except ValueError:
+                # Then it may just be a string, we cannot do anything with it
+                logger.debug('Failed to parse anchore_error_json as json')
+                return ret
+
+            if {'message', 'httpcode', 'detail'}.issubset(set(err_json)):
+                ret.update(err_json)
 
             try:
-                if {'error_code'}.issubset(set(errmsg.__dict__['anchore_error_json'])) and errmsg.__dict__['anchore_error_json'].get('error_code', None):
+                if {'error_code'}.issubset(set(err_json)) and err_json.get('error_code', None):
                     if 'error_codes' not in ret['detail']:
                         ret['detail']['error_codes'] = []
-                    ret['detail']['error_codes'].append(errmsg.__dict__['anchore_error_json'].get('error_code'))
-            except Exception as err:
-                try:
-                    logger.warn("unable to marshal error details: source error {}".format(errmsg.__dict__))
-                except:
-                    pass
+                    ret['detail']['error_codes'].append(err_json.get('error_code'))
+            except KeyError:
+                logger.warn("unable to marshal error details: source error {}".format(errmsg.__dict__))
     return ret
 
 

--- a/anchore_engine/services/catalog/api/controllers/default_controller.py
+++ b/anchore_engine/services/catalog/api/controllers/default_controller.py
@@ -4,6 +4,7 @@ import anchore_engine.apis
 from anchore_engine import db
 import anchore_engine.services.catalog.catalog_impl
 import anchore_engine.common
+from anchore_engine.services.catalog.archiver import ImageConflict
 from anchore_engine.subsys import logger
 import anchore_engine.configuration.localconfig
 import anchore_engine.subsys.servicestatus
@@ -95,7 +96,9 @@ def add_image(image_metadata=None, tag=None, digest=None, created_at=None, from_
 
             with db.session_scope() as session:
                 return_object, httpcode = anchore_engine.services.catalog.catalog_impl.image(session, request_inputs, bodycontent=image_metadata)
-
+    except ImageConflict as img_err:
+        httpcode = 409
+        return_object = str(img_err)
     except Exception as err:
         logger.exception('Error processing image add')
         httpcode = 500

--- a/anchore_engine/services/catalog/archiver.py
+++ b/anchore_engine/services/catalog/archiver.py
@@ -79,6 +79,20 @@ class MetadataNotFound(Exception):
     pass
 
 
+class ImageConflict(Exception):
+
+    def __init__(self, msg, detail=None):
+        self.msg = msg
+        self.detail = detail
+
+    def __str__(self):
+        return str(json.dumps({
+            'httpcode': 409,
+            'message': self.msg,
+            'detail': self.detail
+        }))
+
+
 class ObjectStoreLocation(JsonMappedMixin):
     class ObjectStoreLocationV1Schema(JitSchema):
         bucket = fields.Str()
@@ -716,7 +730,7 @@ class RestoreArchivedImageTask(object):
         with session_scope() as session:
             if db_catalog_image.get(self.image_digest, self.account, session):
                 logger.info('Image archive restore found existing image records already. Aborting restore.')
-                raise Exception('Conflict: Image already exists in system. No restore possible')
+                raise ImageConflict('Conflict: Image already exists in system. No restore possible')
 
             rec = db_archived_images.get(session, self.account, self.image_digest)
             if not rec:
@@ -844,7 +858,7 @@ class RestoreArchivedImageTaskFromArchiveTarfile(RestoreArchivedImageTask):
         with session_scope() as session:
             if db_catalog_image.get(self.image_digest, self.account, session):
                 logger.info('Image archive restore found existing image records already. Aborting restore.')
-                raise Exception('Conflict: Image already exists in system. No restore possible')
+                raise ImageConflict('Conflict: Image already exists in system. No restore possible')
 
         dest_obj_mgr = object_store.get_manager()
 

